pd/src/app.rs

```diff
@@ -335,29 +333,106 @@ impl App {
+                    tracing::debug!(?current_rate);
+                }
+
+                // here's where we should compute updated rates (moved from state code)
+                // TODO @ava insert calls here
+
+                // move below into calls to methods in stake crate?
+
+                // 1 bps = 1e-4, so here we group digits by 4s rather than 3s as is usual
+                let base_reward_rate = 3_0000; // 3bps -> 11% return over 365 epochs, why not
+                let base_exchange_rate = (current_base_rate.base_exchange_rate
+                    * (base_reward_rate + 1_0000_0000))
+                    / 1_0000_0000;
+                let next_base_rate = BaseRateData {
+                    base_exchange_rate,
+                    base_reward_rate,
+                    epoch_index: epoch.index + 1,
+                };
+
+                let next_rates = current_rates
+                    .iter()
+                    .map(|current_rate| {
+                        // TODO (hdevalence) use funding streams here, this ignores funding streams
+                        let validator_reward_rate = base_reward_rate; // (minus fs term)
+                        let validator_exchange_rate = (current_rate.validator_exchange_rate
+                            * (validator_reward_rate + 1_0000_0000))
+                            / 1_0000_0000;
+
+                        // this is supposed to be multiplied by the number of delegation tokens,
+                        // how do we track that?
+                        let _voting_power_adjustment =
+                            (validator_exchange_rate * 1_0000_0000) / base_exchange_rate;
+
+                        RateData {
+                            identity_key: current_rate.identity_key,
+                            epoch_index: epoch.index + 1,
+                            // TODO: update this as we track the delegations
+                            voting_power: current_rate.voting_power,
+                            validator_exchange_rate,
+                            validator_reward_rate,
+                        }
+                    })
+                    .collect();
+
+                tracing::debug!(?next_base_rate);
+                for next_rate in &next_rates {
+                    tracing::debug!(?next_rate);
+                }
+
+                pending_block.lock().unwrap().next_rates = Some(next_rates);
+                pending_block.lock().unwrap().next_base_rate = Some(BaseRateData {
+                    epoch_index: epoch.index + 1,
+                    base_reward_rate,
+                    base_exchange_rate,
+                });
+            }
+
+            // TODO: if necessary, set the EndBlock response to add validators
+            // at the epoch boundary
+            if end.height.unsigned_abs() == epoch.start_height().value() {
+                // TODO: does this go in the first block of the epoch or the last block of the epoch?
```

Comment:
Should we be updating the consensus set in the first block of each epoch, or the last block of each epoch? I think we might want to be doing it in the last block of epoch e, so that the first block of epoch e+1 will be voted on by the new validators.

Reply:
That sounds like the least-confusing thing to me.